File: app/src/main/java/com/example/padamlight/ui/map/fragment/MapFragment.java
package com.example.padamlight.ui.map.fragment;

import android.content.Context;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.DrawableRes;
import android.support.annotation.Nullable;
import android.util.Log;
import android.widget.Toast;


import com.example.padamlight.R;

import com.example.padamlight.data.local.Route;
import com.example.padamlight.data.remote.dto.Direction;
import com.example.padamlight.data.remote.dto.DirectionRoutesLegs;
import com.example.padamlight.data.remote.services.GoogleMapProvider;
import com.example.padamlight.enums.MarkerType;
import com.example.padamlight.data.local.Suggestion;
import com.example.padamlight.ui.base.BaseFragment;
import com.example.padamlight.ui.map.interfaces.MapActionsDelegate;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.BitmapDescriptor;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.maps.model.Polyline;
import com.google.android.gms.maps.model.PolylineOptions;
import com.google.maps.android.PolyUtil;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;



/**
 * Map Fragment
 * Responsible of displaying map and interactions with it
 */
public class MapFragment extends BaseFragment implements OnMapReadyCallback, MapActionsDelegate {


    @Nullable
    private GoogleMap mMap;

    private GoogleMapProvider googleMapProvider;

    public MapFragment() {
        super(R.layout.fragment_map);

    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        googleMapProvider = new GoogleMapProvider();


    }

    @Override
    public void initUi() {
        // Instanciate map fragment
        SupportMapFragment mapFragment = (SupportMapFragment) getChildFragmentManager().findFragmentById(R.id.map_fragment);
        if (mapFragment != null) {
            mapFragment.getMapAsync(this);
        }

    }


    public static MapFragment newInstance() {
        return new MapFragment();
    }


    @Override
    public void onMapReady(GoogleMap googleMap) {
        mMap = googleMap;
        if (mMap != null) {
            mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(new LatLng(47.902964, 1.9092510000000402), 16f));
        }
    }

    @Override
    public void updateMap(LatLng... latLngs) {
        if (mMap != null) {
            LatLngBounds.Builder builder = new LatLngBounds.Builder();
            for (LatLng latLng : latLngs) {
                builder.include(latLng);
            }
            LatLngBounds bounds = builder.build();
            animateMapCamera(bounds);
        }
    }

    @Override
    public void updateMarker(MarkerType type, String markerName, LatLng markerLatLng) {
        if (mMap != null) {
            MarkerOptions marker = new MarkerOptions()
                    .position(markerLatLng)
                    .title(markerName)
                    .icon(getMarkerIcon(type));

            mMap.addMarker(marker);
        }
    }

    @Override
    public void clearMap() {
        if (mMap != null) {
            mMap.clear();
        }
    }

    @Override
    public void drawRoute(final Suggestion from, final Suggestion to) {

        Call<Direction> response = googleMapProvider.getDirections(from.toString(), to.toString(), getString(R.string.google_maps_key));

        response.enqueue(new Callback<Direction>() {
            @Override
            public void onResponse(Call<Direction> call, Response<Direction> response) {
                if(response.body() != null){

                    Direction direction = response.body();

                    if(direction.getRoutes().length > 0){
                        DirectionRoutesLegs legs = direction.getRoutes()[0].getLegs()[0];
                        Route route = new Route(from.toString(), to.toString(),
                                legs.getStart_location().getLat(), legs.getStart_location().getLng(),
                                legs.getEnd_location().getLat(), legs.getEnd_location().getLng(),
                                direction.getRoutes()[0].getOverview_polyline().getPoints());

                        PolylineOptions polylineOptions = setupPolylines(getContext());
                        List<LatLng> pointsList = PolyUtil.decode(route.getOverviewPolyline());
                        for(LatLng point : pointsList){
                            polylineOptions.add(point);
                        }
                        Polyline currentPolyline = mMap.addPolyline(polylineOptions);
                        updateMap(from.getLatLng(), to.getLatLng());

                    }
                }

            }

            @Override
            public void onFailure(Call<Direction> call, Throwable t) {
                Toast.makeText(getContext(), "IMPOSSIBLE DE DESSINER LA ROUTE", Toast.LENGTH_LONG);
            }
        });
    }

    public PolylineOptions setupPolylines(Context context){
        PolylineOptions polylineOptions = new PolylineOptions();

        polylineOptions.width(20);
        polylineOptions.geodesic(true);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            polylineOptions.color(context.getResources().getColor(R.color.colorAccent, context.getTheme()));
        }else{
            polylineOptions.color(context.getResources().getColor(R.color.colorAccent));
        }

        return polylineOptions;

    }

    private void animateMapCamera(LatLngBounds bounds) {
        if (mMap != null) {
            mMap.animateCamera(CameraUpdateFactory.newLatLngBounds(bounds, 100));
        }
    }

    private BitmapDescriptor getMarkerIcon(MarkerType type) {
        @DrawableRes int icon = R.drawable.ic_pickup;

        switch (type) {
            case PICKUP: {
                icon = R.drawable.ic_pickup;
                break;
            }
            case DROPOFF: {
                icon = R.drawable.ic_dropoff;
                break;
            }
        }
        return BitmapDescriptorFactory.fromResource(icon);
    }


}

File: app/src/main/java/com/example/padamlight/ui/map/fragment/MapFragment.kt
package com.example.padamlight.ui.map.fragment


import android.content.Context
import android.os.Build
import android.os.Bundle
import android.support.annotation.DrawableRes
import android.widget.Toast
import com.example.padamlight.R
import com.example.padamlight.data.local.Route
import com.example.padamlight.data.local.Suggestion
import com.example.padamlight.data.remote.dto.Direction
import com.example.padamlight.data.remote.services.GoogleMapProvider
import com.example.padamlight.enums.MarkerType
import com.example.padamlight.ui.base.BaseFragment
import com.example.padamlight.ui.map.interfaces.MapActionsDelegate
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.*
import com.google.maps.android.PolyUtil
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response


/**
 * Map Fragment
 * Responsible of displaying map and interactions with it
 */
class MapFragment : BaseFragment(R.layout.fragment_map), OnMapReadyCallback, MapActionsDelegate {

    companion object {
        fun newInstance(): MapFragment {
            return MapFragment()
        }
    }


    private var mMap: GoogleMap? = null

    private var googleMapProvider: GoogleMapProvider? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        googleMapProvider = GoogleMapProvider()


    }

    override fun initUi() {
        // Instanciate map fragment
        val mapFragment = childFragmentManager.findFragmentById(R.id.map_fragment) as SupportMapFragment?
        mapFragment?.getMapAsync(this)

    }


    override fun onMapReady(googleMap: GoogleMap) {
        mMap = googleMap
        if (mMap != null) {
            mMap!!.moveCamera(CameraUpdateFactory.newLatLngZoom(LatLng(47.902964, 1.9092510000000402), 16f))
        }
    }

    override fun updateMap(vararg latLngs: LatLng) {
        if (mMap != null) {
            val builder = LatLngBounds.Builder()
            for (latLng in latLngs) {
                builder.include(latLng)
            }
            val bounds = builder.build()
            animateMapCamera(bounds)
        }
    }

    override fun updateMarker(type: MarkerType, markerName: String, markerLatLng: LatLng) {
        if (mMap != null) {
            val marker = MarkerOptions()
                    .position(markerLatLng)
                    .title(markerName)
                    .icon(getMarkerIcon(type))

            mMap!!.addMarker(marker)
        }
    }

    override fun clearMap() {
        if (mMap != null) {
            mMap!!.clear()
        }
    }

    override fun drawRoute(from: Suggestion, to: Suggestion) {

        val response = googleMapProvider?.getDirections(from.toString(), to.toString(), getString(R.string.google_maps_key))

        response?.enqueue(object : Callback<Direction> {
            override fun onResponse(call: Call<Direction>, response: Response<Direction>) {
                if (response.body() != null) {

                    val direction = response.body()

                    if (direction?.routes!!.isNotEmpty()) {
                        val legs = direction.routes!![0].legs!![0]
                        val route = Route(from.toString(), to.toString(),
                                legs.start_location!!.lat, legs.start_location!!.lng,
                                legs.end_location!!.lat, legs.end_location!!.lng,
                                direction.routes!![0].overview_polyline!!.points!!)

                        val polylineOptions = setupPolylines(context)
                        val pointsList = PolyUtil.decode(route.overviewPolyline)
                        for (point in pointsList) {
                            polylineOptions.add(point)
                        }
                        mMap!!.addPolyline(polylineOptions)
                        updateMap(from.latLng, to.latLng)

                    }
                }

            }

            override fun onFailure(call: Call<Direction>, t: Throwable) {
                Toast.makeText(context, R.string.no_road, Toast.LENGTH_LONG).show()
            }
        })
    }

    fun setupPolylines(context: Context?): PolylineOptions {
        val polylineOptions = PolylineOptions()

        polylineOptions.width(20f)
        polylineOptions.geodesic(true)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            polylineOptions.color(context!!.resources.getColor(R.color.colorAccent, context.theme))
        } else {
            polylineOptions.color(context!!.resources.getColor(R.color.colorAccent))
        }

        return polylineOptions

    }

    private fun animateMapCamera(bounds: LatLngBounds) {
        if (mMap != null) {
            mMap!!.animateCamera(CameraUpdateFactory.newLatLngBounds(bounds, 100))
        }
    }

    private fun getMarkerIcon(type: MarkerType): BitmapDescriptor {
        @DrawableRes var icon: Int

        icon = when (type) {
            MarkerType.PICKUP -> {
                R.drawable.ic_pickup
            }
            MarkerType.DROPOFF -> {
                R.drawable.ic_dropoff
            }
        }
        return BitmapDescriptorFactory.fromResource(icon)
    }



}


